crypto-sodium/lib/Crypto/Sodium/Salt.hs

```diff
@@ -0,0 +1,57 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+-- SPDX-FileCopyrightText: 2021 Serokell
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+-- | This module gives different ways of obtaining salts.
+--
+-- A “salt” is additional input provided to certain functions.
+-- Unlike a nonce, salt is used for “namespacing”, so there is not
+-- requirement that it is used only once, instead you use different
+-- salts to deisgnate different applications of an algorithm.
+--
+-- This is a subtle and pure semantical difference, so, for convenience,
+-- we also re-export some functions from "Crypto.Sodium.Nonce".
+module Crypto.Sodium.Salt
+  (
+  -- * Literals
+    utf8Lit
+
+  -- * Random salt generation
+  , Crypto.Sodium.Nonce.generate
+  ) where
+
+import Data.ByteArray.Sized (SizedByteArray, sizedByteArray)
+import Data.ByteString (ByteString)
+import Data.Proxy (Proxy (Proxy))
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import GHC.TypeLits (KnownNat, KnownSymbol, symbolVal)
+
+import qualified Crypto.Sodium.Nonce
+
+
+-- | Make a /sized/ 'ByteString' from a type-level string literal.
+--
+-- @
+-- case utf8Lit \@"hello" \@5 of
+--   Just bytes -> {- ... -}  -- bytes :: 'SizedByteArray' 5 ByteString
+--   Nothing -> {- the size did not match the expected one -}
+-- @
+--
+-- Unfortunately, it is impossible to know the length of a UTF8-encoded
+-- string at type-level, so you have to explicitly provide the right length
+-- and the function will return 'Nothing' if you fail to do so.
+--
+-- You have to provide the string as a type-level symbol rather than a
+-- value-level string to make sure it is harder to pass here something
+-- that is not really a literal.
+utf8Lit
+  :: forall s n.
+     (KnownSymbol s, KnownNat n)
+  => Maybe (SizedByteArray n ByteString)
+utf8Lit = sizedByteArray $ T.encodeUtf8 (T.pack str)
```

Comment:
It's really unclear whether this function is useful at all. It's basically `sizedByteArray @n "string"` sans utf-8 (which might not even be the right encoding for a particular application). And `Maybe` is rather pointless, since `Nothing` branch is invariably bottom, at least in all use cases I could think of. I don't know if we want to include this, frankly, it feels very awkward and somewhat pointless.

Wrt type-level length, turns out it's actually possible to hack something together, at least for the printable subset of ASCII, see e.g. https://blog.csongor.co.uk/symbol-parsing-haskell/, but at the end of the day I don't think the required effort is worth it.

Another option is to provide a TH function that would just paste the required length. Something like this, perhaps:

```haskell
sizedBytes bytes = [| unsafeSizedByteArray @$n bytes |]
  where n = LitT . NumTyLit $ length bytes
```

(not sure if this syntax works from the top of my head, but if not, it's easy enough to manually construct `AppTypeE` node)

I don't know if TH makes much sense either, but it seems like ultimately a less awkward option.

Reply:
I think I really like the TH idea 🤔. I’ll give it a go (I wanted to learn TH anyway).